extensions/amp-lightbox/1.0/amp-lightbox.js

```diff
@@ -50,31 +50,36 @@ class AmpLightbox extends PreactBaseElement {
     this.registerApiAction('open', (api) => api.open(), ActionTrust.LOW);
     this.registerApiAction('close', (api) => api.close(), ActionTrust.LOW);
     return dict({
-      'onBeforeOpen': this.beforeOpen_.bind(this),
-      'onAfterClose': this.afterClose_.bind(this),
+      'onBeforeOpen': this.toggle_.bind(this, true),
+      'onAfterClose': this.toggle_.bind(this, false),
     });
   }
 
-  /**
-   * Setting hidden to false
-   * @private
-   */
-  beforeOpen_() {
-    this.open_ = true;
-    toggle(this.element, true);
-    this.element.setAttribute('open', '');
-    fireEvent(this.win, this.element, 'open', ActionTrust.HIGH);
+  /** @override */
+  triggerEvent(element, eventName, detail) {
+    const event = createCustomEvent(
+      toWin(element.ownerDocument.defaultView),
+      `amp-lightbox.${eventName}`,
+      detail
+    );
+    Services.actionServiceForDoc(element).trigger(
+      element,
+      eventName,
+      event,
+      ActionTrust.HIGH
+    );
+
+    super.triggerEvent(element, eventName, detail);
   }
 
   /**
-   * Setting hidden to true
-   * @private
+   * Toggle open/closed attributes.
+   * @param {boolean} opt_state
    */
-  afterClose_() {
-    this.open_ = false;
-    toggle(this.element, false);
-    this.element.removeAttribute('open');
-    fireEvent(this.win, this.element, 'close', ActionTrust.HIGH);
+  toggle_(opt_state) {
```

Comment:
Is toggle necessary outside of AMP mode? If so, this should probably be on `BaseElement`

Reply:
AFAICT, the Lightbox Preact component `component.js` uses its own `setVisible`/`setStyle` states to manage visibility, and I interpreted this to mean that the toggle functionality is only being provided for the API/imperative handle stuff to bridge the Preact-AMP gap. @caroqliu , can you confirm if this is the case?

Reply:
This is needed in both modes since the `hidden` attr is used in CSS and applying the `open` attr is generally valuable to have in all HTML contexts.

One note about the visibility state management in `component.js` is that this handles internals such as animations, but needs to coordinate with HTML layer so the animations can even be seen.

Reply:
Addressed in latest commit https://github.com/ampproject/amphtml/pull/32704/commits/840624b482234f952372106f643570d00f4dfe05